lib/ui/platform_dispatcher.dart

```diff
@@ -1299,6 +1303,102 @@ class PlatformDispatcher {
+  /// so text in the app resizes properly when the text scaling preference
+  /// changes.
+  double? scaleFontSize(double unscaledFontSize) {
+    assert(unscaledFontSize >= 0);
+    assert(unscaledFontSize.isFinite);
+
+    if (textScaleFactor == 1.0) {
+      return unscaledFontSize;
+    }
+
+    final int unscaledFloor = unscaledFontSize.floor();
+    final int unscaledCeil = unscaledFontSize.ceil();
+    if (unscaledFloor == unscaledCeil) {
+      return _scaleAndMemoize(unscaledFloor);
+    }
+    assert(unscaledCeil - unscaledFloor == 1);
+
+    return switch ((_scaleAndMemoize(unscaledFloor), _scaleAndMemoize(unscaledCeil))) {
+      (null, _) => null,
+      (_, null) => null,
+      (final double lower, final double upper) => lower + (upper - lower) * (unscaledFontSize - unscaledFloor),
+    };
+  }
+
+  // The _cachedFontSizes variable is set to null if GetScaledFontSize is not
+  // implemented on the current platform.
+  Map<int, double>? _cachedFontSizes = <int, double>{};
+  double? _scaleAndMemoize(int unscaledFontSize) {
+    final Map<int, double>? cache = _cachedFontSizes;
+    final int? configurationGenerationId = _configuration.configurationId;
+    if (cache == null || configurationGenerationId == null) {
+      return null;
+    }
+    final double? cachedValue = cache[unscaledFontSize];
+    if (cachedValue != null) {
+      assert(cachedValue >= 0);
+      return cachedValue;
+    }
+
+    final double unscaledFontSizeInt = unscaledFontSize.toDouble();
+    final double fontSize = PlatformDispatcher._getScaledFontSize(unscaledFontSizeInt, configurationGenerationId);
+    switch (fontSize) {
+      case >= 0:
+        return cache.putIfAbsent(unscaledFontSize, () => fontSize);
+      case -1:
+        assert(cache.isEmpty);
+        return _cachedFontSizes = null;
+      case -2:
+        assert(false, 'Flutter Error: incorrect configuration id.');
+        return null;
+      case final double errorCode:
+        assert(false, 'GetScaledFontSize failed with $errorCode');
+        return null;
+    }
+  }
+
+  // Scales the `unscaledFontSize` using the platform's text scaling curve.
+  //
+  // Currently this is only implemented on newer versions of Android (SDK level
+  // 34, using the `TypedValue#applyDimension` API). This function returns -1 on
+  // other platforms.
+  //
+  // The `configurationGeneration` parameter tells the embedder which platform
+  // configuration to use for computing the scaled font size. When the user
+  // changes the platform configuration, the configuration data will first be
+  // made available on the platform thread before being dispatched asynchronously
+  // to the Flutter UI thread. Since this call is synchronous, it could be using
+  // the newer configuration that has yet been sent to Flutter. This call should
+  // use the [PlatformDispatcher]'s latest configuration generation to avoid such
+  // a race.
```

Comment:
This should document that you can only use the last generation known by the PlatformDispatcher. You cannot use any arbitrary old generations. (which I think makes "generation" a better name for this then the generic "ID").